src/include/storage/local_storage/local_hash_index.h

```diff
@@ -56,45 +62,74 @@ class HashIndexLocalStorage final : public BaseHashIndexLocalStorage {
     }
 
     void reserveSpaceForAppend(uint32_t numNewEntries) {
+        std::unique_lock xLock{mtx};
+        reserveSpaceForAppendNoLock(numNewEntries);
+    }
+
+    void reserveSpaceForAppendNoLock(uint32_t numNewEntries) {
         localInsertions.reserveSpaceForAppend(numNewEntries);
     }
 
     bool append(OwnedType&& key, common::offset_t value, visible_func isVisible) {
+        std::unique_lock xLock{mtx};
+        return appendNoLock(std::move(key), value, isVisible);
+    }
+
+    bool appendNoLock(OwnedType&& key, common::offset_t value, visible_func isVisible) {
         return localInsertions.append(std::move(key), value, isVisible);
     }
 
     size_t append(IndexBuffer<OwnedType>& buffer, uint64_t bufferOffset, visible_func isVisible) {
+        std::unique_lock xLock{mtx};
+        return appendNoLock(buffer, bufferOffset, isVisible);
+    }
+
+    size_t appendNoLock(IndexBuffer<OwnedType>& buffer, uint64_t bufferOffset,
+        visible_func isVisible) {
         return localInsertions.append(buffer, bufferOffset, isVisible);
     }
 
-    bool hasUpdates() const { return !(localInsertions.empty() && localDeletions.empty()); }
+    bool hasUpdates() {
+        std::shared_lock sLock{mtx};
+        return !(localInsertions.empty() && localDeletions.empty());
+    }
 
-    int64_t getNetInserts() const {
+    int64_t getNetInserts() {
+        std::shared_lock sLock{mtx};
         return static_cast<int64_t>(localInsertions.size()) - localDeletions.size();
     }
 
     void clear() {
+        std::unique_lock xLock{mtx};
         localInsertions.clear();
         localDeletions.clear();
     }
 
     void applyLocalChanges(const std::function<void(KeyType)>& deleteOp,
         const std::function<void(const InMemHashIndex<T>&)>& insertOp) {
+        std::shared_lock sLock{mtx};
         for (auto& key : localDeletions) {
             deleteOp(key);
         }
         insertOp(localInsertions);
     }
 
-    void reserveInserts(uint64_t newEntries) { localInsertions.reserve(newEntries); }
+    void reserveInserts(uint64_t newEntries) {
+        std::unique_lock xLock{mtx};
+        localInsertions.reserve(newEntries);
```

Comment:
I don't think this is the right behaviour here (though it should be safe). If we have multiple threads bulk reserving we should reserve the combined amount.
This also applies to `reserveSpaceForAppend`, and maybe the solution should be to track capacity explicitly and merge the two functions, and in both situations just add to the capacity (even now, I think `reserveInserts` could be replaced by `reserveSpaceForAppend` since I think `reserveInserts` is only called when the number of existing entries is zero). I think that should be safe, since the local storage gets cleared during rollbacks, so if there is an interruption we won't just keep growing the capacity if the reserve is tried again.